### ISSUE TITLE
Arbitrary math parsing

### ISSUE THREAD

**Issue by ddavidebor** [elastic/timelion#174]
I wanted to ask the opinion of the project managers:

Writing from scratch all the math functions is slow. I suggest using a more advanced solution such as [Math.js](http://mathjs.org/examples/chaining.js.html) that allows chaining of functions, [expression parsing](http://mathjs.org/examples/expressions.js.html), can efficiently manage matrix of data and convert between [units](http://mathjs.org/examples/units.js.html).

There is no need to reinvent the wheel, most of the statistical functions are already [present](http://mathjs.org/docs/reference/functions.html)

Also, the possibility to print in latex and debug with understandable errors ([example](http://mathjs.org/examples/browser/pretty_printing_with_mathjax.html)) would solve #157  and allow an easy-to-understand visualization for people who have a background more into engineering/physics (like myself) than software.

Timelion can do computations server-side right?
Davide
Fermium LABS srl

**Comment by rashidkpc** [comment]
I'd take a pull for adding a Math.js function, however that might look.

Timelion does all computations server side.

**Comment by ddavidebor** [comment]
@rashidkpc we're working on an implementation as a plugin, using your [timelion-random](https://github.com/rashidkpc/timelion-random) as a basis.

I'll close the issue so it does not clutter, but here are a few reasonable features:

* full math functions with syntax such as `.es(query).math("this*2")`
* treat with units conversion `.es(query).math("this to Kw")`
* save variables `.es(current).math(i=this).hide(), .es(voltage).assign(v=this).hide()`
* retrieve and elaborate on variables `.math(power=(v*i))` or `.math(v*i).label(power)`
* very fast vector math
* thernary conditions, various test operators
* string to number
* scientific costants, trigonometry and all sort of stuff
* easily import other statistical and math functions in a very similar way to the one Timelion uses, but instead allowing them to be used inside math equations

Other function that may be developed in order to make this usable, thought not now:

* multiline editor, such as in issue #31 
* Print the equations to latex and render in realtime

We're working [here](https://github.com/fermiumlabs/mathlion)

Davide

**Comment by rashidkpc** [comment]
That looks so cool!
